File: pattern/GoF/behavioral/strategy/strategy.go
package main

import "fmt"

type PricingStrategy interface {
	CalculatePrice(basePrice float64) float64
}

type RegularPricing struct{}

func (rp RegularPricing) CalculatePrice(basePrice float64) float64 {
	return basePrice
}

type StudentPricing struct{}

func (sp StudentPricing) CalculatePrice(basePrice float64) float64 {
	return basePrice * 0.8
}

type SeniorPricing struct{}

func (sp SeniorPricing) CalculatePrice(basePrice float64) float64 {
	return basePrice * 0.7
}

type VIPPricing struct{}

func (vp VIPPricing) CalculatePrice(basePrice float64) float64 {
	return basePrice * 1.2
}

type TravelPackage struct {
	name            string
	basePrice       float64
	pricingStrategy PricingStrategy
}

func NewTravelPackage(name string, basePrice float64, pricingStrategy PricingStrategy) *TravelPackage {
	return &TravelPackage{name: name, basePrice: basePrice, pricingStrategy: pricingStrategy}
}

func (tp TravelPackage) GetPrice() float64 {
	return tp.pricingStrategy.CalculatePrice(tp.basePrice)
}

func (tp *TravelPackage) SetPricingStrategy(ps PricingStrategy) {
	tp.pricingStrategy = ps
}

func main() {
	regularPricing := RegularPricing{}
	studentPricing := StudentPricing{}
	seniorPricing := SeniorPricing{}
	vipPricing := VIPPricing{}

	tokyoPackage := NewTravelPackage("Tokyo Adventure", 1000, regularPricing)

	fmt.Println("Regular price:", tokyoPackage.GetPrice())

	tokyoPackage.SetPricingStrategy(studentPricing)
	fmt.Println("Student price:", tokyoPackage.GetPrice())

	tokyoPackage.SetPricingStrategy(seniorPricing)
	fmt.Println("Senior price:", tokyoPackage.GetPrice())

	tokyoPackage.SetPricingStrategy(vipPricing)
	fmt.Println("VIP price:", tokyoPackage.GetPrice())
}

File: pattern/GoF/behavioral/strategy/strategy.rs
trait PricingStrategy {
    fn calculate_price(&self, base_price: f64) -> f64;
}

struct RegularPricing;
impl PricingStrategy for RegularPricing {
    fn calculate_price(&self, base_price: f64) -> f64 {
        base_price
    }
}

struct StudentPricing;
impl PricingStrategy for StudentPricing {
    fn calculate_price(&self, base_price: f64) -> f64 {
        base_price * 0.8
    }
}

struct SeniorPricing;
impl PricingStrategy for SeniorPricing {
    fn calculate_price(&self, base_price: f64) -> f64 {
        base_price * 0.7
    }
}

struct VIPPricing;
impl PricingStrategy for VIPPricing {
    fn calculate_price(&self, base_price: f64) -> f64 {
        base_price * 1.2
    }
}

struct TravelPackage {
    name: String,
    base_price: f64,
    // dynamic dispatch を行うため Box<dyn PricingStrategy> として定義
    pricing_strategy: Box<dyn PricingStrategy>,
}

impl TravelPackage {
    fn new(name: String, base_price: f64, pricing_strategy: Box<dyn PricingStrategy>) -> Self {
        Self {
            name,
            base_price,
            pricing_strategy,
        }
    }

    fn get_price(&self) -> f64 {
        self.pricing_strategy.calculate_price(self.base_price)
    }

    fn set_pricing_strategy(&mut self, pricing_strategy: Box<dyn PricingStrategy>) {
        self.pricing_strategy = pricing_strategy;
    }
}

fn main() {
    let mut tokyo_package = TravelPackage::new(
        String::from("Tokyo Adventure"),
        1000.0,
        Box::new(RegularPricing),
    );
    println!("Package: {}", tokyo_package.name);

    println!("Regular price: ${:.2}", tokyo_package.get_price());

    tokyo_package.set_pricing_strategy(Box::new(StudentPricing));
    println!("Student price: ${:.2}", tokyo_package.get_price());

    tokyo_package.set_pricing_strategy(Box::new(SeniorPricing));
    println!("Senior price: ${:.2}", tokyo_package.get_price());

    tokyo_package.set_pricing_strategy(Box::new(VIPPricing));
    println!("VIP price: ${:.2}", tokyo_package.get_price());
}
